labcontrol/db/process.py

```diff
@@ -3819,3 +3826,407 @@ def _generate_amplicon_prep_information(self):
+                id = d['extraction_kit_id']
+                d['extraction_kit_lot'] = reagent[id]['external_lot_id']
+
+                id = d['epmotion_tool_id']
+                d['epmotion_tool_name'] = equipment[id]['external_id']
+
+                # for now, platform is hard-coded to 'Illumina'
+                # will need to change once Nanopore is supported by LC
+                # and we have a column to record one or the other.
+                d['platform'] = 'Illumina'
+
+                # these key/value pairs are tentatively hard-coded for now.
+                d['sequencing_method'] = 'sequencing by synthesis'
+                d['run_center'] = 'UCSDMI'
+                d['library_construction_protocol'] = 'KL KHP'
+
+                # EXPERIMENT_DESIGN_DESCRIPTION as with Amplicon, will remain
+                # empty when NULL.
+
+                # Replicating logic from Amplicon pre-processing
+                # TODO: refactor to a shared method
+                d['orig_name2'] = d['orig_name']
+
+                # Note is_control field replaces the line directly below
+                if not d['is_control']:
+                    # strip the prepended study id from orig_name2, but only
+                    # if this is an 'experimental sample' row, and not a
+                    # 'control' row. (captured here w/orig_name2 and study_id
+                    # not equaling None. This also prevents interference w/the
+                    # population of the DataFrame index below, using the
+                    # existing list comprehension.
+                    d['orig_name2'] = re.sub("^%s\." % d['study_id'],
+                                             '',
+                                             d['orig_name2'])
+
+            return results
+
+    def _generate_metagenomics_prep_information(self):
+        """Generates prep information for Metagenomics workflows
+
+        An internal method used to implement the generation of prep information
+        files for Metagenomics workflows. This method is called by
+        generate_prep_information() only. This method is not currently
+        implemented.
+
+        Returns
+        -------
+        dict: { int: str,
+                int: str,
+                int: str,
+                .
+                .
+                .
+                int: str,
+                str: str }
+
+        where 'int: str' represents either a Study ID and a TSV file (in string
+        form), or a Prep ID and TSV file (in string form).
+
+        'str: str' represents controls data; the key is the constant
+        'Controls', and the value is a TSV file (in string form).
+        """
+        results = self._get_metagenomics_data_for_prep()
+
+        data = {}
+
+        for item in results:
+            # format well
+            well = []
+            col = item['col_num']
+            row = item['row_num']
+            while row:
+                row, rem = divmod(row - 1, 26)
```

Comment:
Weird things I believe would happen if `row` were `<= 0`. The database does not appear to enforce that the row value is `[1, inf)`:

https://github.com/jdereus/labman/blob/313d006ccad1d9111d218b7ce645eb6d10fe37a9/labcontrol/db/support_files/db_patch.sql#L242-L249

Would it be possible to place an assertion on the valid range for both `row` and `col` out of an abundance of caution?

Reply:
...or should constraints be added elsewhere?

Reply:
Might want to consider logging this as a future issue rather than changing it here, because this very same logic exists in `_generate_amplicon_prep_information` and should probably be centralized before we start fixing it :)

Reply:
Short of changing the datatypes to a user-defined (domain) type, we could add constraints like these:
ALTER TABLE labman.well ADD CONSTRAINT rowchk CHECK (row_num > 0);
ALTER TABLE labman.well ADD CONSTRAINT colchk CHECK (col_num > 0);